sql/src/main/java/io/crate/operation/collect/BatchIteratorCollectorBridge.java

```diff
@@ -48,30 +48,42 @@ public static CrateCollector newInstance(Supplier<CompletableFuture<BatchIterato
         private final BatchIterator batchIterator;
         private final BatchConsumer consumer;
 
-        private boolean started;
+        private boolean running;
 
         SyncBatchItCollector(BatchIterator batchIterator, BatchConsumer consumer) {
             this.batchIterator = batchIterator;
             this.consumer = consumer;
         }
 
         @Override
-        public synchronized void doCollect() {
-            if (!started) {
-                started = true;
+        public void doCollect() {
+            boolean invokeConsumer = isNotRunning();
+            if (invokeConsumer) {
                 consumer.accept(batchIterator, null);
             }
             // else: got killed
         }
 
         @Override
-        public synchronized void kill(Throwable throwable) {
-            if (started) {
-                batchIterator.kill(throwable);
-            } else {
-                started = true;
+        public void kill(Throwable throwable) {
+            boolean invokeConsumer = isNotRunning();
+            if (invokeConsumer) {
                 consumer.accept(null, throwable);
+            } else {
+                batchIterator.kill(throwable);
+            }
+        }
+
+        private boolean isNotRunning() {
```

Comment:
This method is extremely confusing as the name implies reading a flag whereas it is also setting running to true. So I get that the intent is to *know* if we currently started running it or if it was *already* running. So how about using an AtomicBoolean.compareAndSet(false, true) and only invoke consumer.accept if this returns true